### PR TITLE
perf(pt2): optimize .pt2 C++ inference path

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary
- Replace CPU-side `buildTypeSortedNlist` with `createNlistTensor(data, expected_nnei)` — avoids distance computation and type sorting every step; model's compiled `format_nlist` handles this on-device
- Export with `do_atomic_virial=False` by default — avoids 3 extra `torch.autograd.grad` backward passes; add `--atomic-virial` flag to `dp convert-backend`
- Cache `mapping_tensor` as member variable, only rebuild when `ago == 0`
- Store `nnei` and `do_atomic_virial` in .pt2 metadata for C++ to read at init
- Make nnei dynamic in torch.export — compiled graph accepts variable-size neighbor lists via internal padding + sort branch

### Before this PR

.pt2 was **9x slower** than .pth due to CPU-side nlist sorting, baked-in atomic virial backward passes, and excessive clones:

| Atoms | .pth (ms/step) | .pt2 (ms/step) | .pt2/.pth |
|------:|---------------:|---------------:|:---------:|
| 192 | 11 | 97 | **8.8x** |

#### DPA1 L0 (se_atten nlayer=0)

| Atoms | .pth (ms) | .pt2 (ms) | .pt2/.pth |
|------:|----------:|----------:|:---------:|
| 192 | 5.60 | 4.93 | **0.88x** |
| 384 | 6.69 | 8.45 | **1.26x** |
| 768 | 10.9 | 16.3 | **1.49x** |
| 1536 | 19.3 | 31.2 | **1.62x** |
| 3072 | 36.7 | 58.8 | **1.60x** |
| 6144 | 72.0 | 116 | **1.62x** |
| 12288 | 140 | 229 | **1.63x** |

#### DPA1 L2 (se_atten nlayer=2)

| Atoms | .pth (ms) | .pt2 (ms) | .pt2/.pth |
|------:|----------:|----------:|:---------:|
| 192 | 13.0 | 9.17 | **0.71x** |
| 384 | 22.2 | 16.2 | **0.73x** |
| 768 | 41.0 | 30.4 | **0.74x** |
| 1536 | 77.8 | 58.8 | **0.76x** |

#### DPA2 (repinit + repformer)

| Atoms | .pth (ms) | .pt2 (ms) | .pt2/.pth |
|------:|----------:|----------:|:---------:|
| 192 | 28.5 | 15.6 | **0.55x** |
| 384 | 34.6 | 28.2 | **0.81x** |
| 768 | 60.5 | 53.4 | **0.88x** |
| 1536 | 112.9 | 104 | **0.92x** |

For models with more compute (DPA1 L2, DPA2), .pt2 is **24-45% faster** than .pth. For the smallest model (DPA1 L0), .pt2 has higher per-call overhead that dominates at large atom counts.

## Test plan
- [x] All Python export/make_fx tests pass (74 tests)
- [x] All Python model tests pass
- [x] All C++ ctest pass (0 failures)
- [x] All 37 LAMMPS .pt2 tests pass
- [x] V100 benchmark confirms speedup for DPA1 L2 and DPA2

## Summary by CodeRabbit

* **New Features**
  * Added `--atomic-virial` command-line flag to enable atomic virial correction during model conversion and export operations
  * Models exported with this feature now include per-atom virial contributions for improved computational accuracy
  * Atomic virial support available for all exportable model formats

### CODE REVIEW

File: source/api_cc/src/commonPTExpt.h [hunk 1]
```diff
@@ -438,100 +438,38 @@ inline std::string read_zip_entry(const std::string& zip_path,
...
-        std::sort(group.begin(), group.end());
-        int count = std::min(static_cast<int>(group.size()), sel[tt]);
-        for (int kk = 0; kk < count; ++kk) {
-          result[offset + col + kk] = group[kk].second;
-        }
-        col += sel[tt];
-      }
+    int ncopy = std::min(static_cast<int>(data[ii].size()), expected_nnei);
+    for (int jj = 0; jj < ncopy; ++jj) {
+      flat_data[static_cast<size_t>(ii) * expected_nnei + jj] = data[ii][jj];
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h]: Fixed in b7509db5 — nlist nnei dimension is now dynamic in the .pt2 export. `createNlistTensor` pads to `nnei+1` (never truncates), so the compiled graph's `_format_nlist` always enters the distance-sort branch on GPU.

File: deepmd/entrypoints/convert_backend.py [hunk 2]
```diff
@@ -20,12 +21,22 @@ def convert_backend(
...
     data = inp_hook(INPUT)
-    out_hook(OUTPUT, data)
+    # Forward atomic_virial to pt_expt deserialize_to_file if applicable
+    import inspect
+
+    sig = inspect.signature(out_hook)
+    if "do_atomic_virial" in sig.parameters:
+        out_hook(OUTPUT, data, do_atomic_virial=atomic_virial)
+    else:
+        out_hook(OUTPUT, data)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/entrypoints/convert_backend.py:46]: Fixed in 217a5876 — `convert_backend` now raises `ValueError` when `--atomic-virial` is passed to a backend whose hook lacks `do_atomic_virial`.

File: source/api_cc/src/commonPTExpt.h [hunk 3]
```diff
@@ -438,100 +438,38 @@ inline std::string read_zip_entry(const std::string& zip_path,
...
     }
   }
-
-  torch::Tensor tensor =
-      torch::from_blob(result.data(), {1, nloc, nsel},
-                       torch::TensorOptions().dtype(torch::kInt64))
-          .clone();
-  return tensor;
+  torch::Tensor flat_tensor = torch::tensor(flat_data, torch::kInt32);
+  return flat_tensor.view({1, nloc, expected_nnei});
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h]: Fixed in b7509db5 — the nlist nnei dimension is now dynamic via `torch.export.Dim("nnei")`. `createNlistTensor` pads to `min_nnei = nnei + 1` without truncation, so the compiled `_format_nlist` always sorts by distance on GPU and truncates to `sum(sel)`.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 4]
```diff
@@ -133,16 +132,29 @@ void DeepSpinPTExpt::init(const std::string& model,
...
+  // Read expected nnei (= sum(sel)) — the .pt2 graph has this dimension static.
+  if (metadata.obj_val.count("nnei")) {
+    nnei = metadata["nnei"].as_int();
+  } else {
+    // Fallback: compute from sel array
+    nnei = 0;
+    for (const auto& v : metadata["sel"].as_array()) {
+      nnei += v.as_int();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:151]: Older .pt2 models are handled by fallback defaults: `do_atomic_virial` defaults to `true` (matching the old export behavior) and `nnei` defaults to `sum(sel)` when the metadata field is absent. There is no versioning scheme in the .pt2 format to do more than this.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 5]
```diff
@@ -398,6 +410,12 @@ void DeepSpinPTExpt::compute(ENERGYVTYPE& ener,
                         fwd_map.size(), nall_real);
 
   if (atomic) {
+    if (!do_atomic_virial) {
+      throw deepmd::deepmd_exception(
+          "Atomic virial is not available in this .pt2 model "
+          "(exported without --atomic-virial). "
+          "Regenerate with: dp convert-backend --atomic-virial INPUT OUTPUT");
+    }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc]: Fixed in 217a5876 — the `do_atomic_virial` check is now before `run_model` in all 4 compute functions (both LAMMPS and standalone paths in DeepPotPTExpt and DeepSpinPTExpt).

[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc]: 217a5876

File: deepmd/entrypoints/convert_backend.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/entrypoints/convert_backend.py:46]: 217a5876

File: source/api_cc/src/commonPTExpt.h [hunk 6]
```diff
@@ -438,100 +438,41 @@ inline std::string read_zip_entry(const std::string& zip_path,
...
   }
-
-  torch::Tensor tensor =
-      torch::from_blob(result.data(), {1, nloc, nsel},
-                       torch::TensorOptions().dtype(torch::kInt64))
-          .clone();
-  return tensor;
+  torch::Tensor flat_tensor = torch::tensor(flat_data, torch::kInt32);
+  return flat_tensor.view({1, nloc, nnei});
 }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h]: Fixed in 862f5609 — `createNlistTensor` moved to `commonPT.h` with a `min_nnei` parameter. All .pt2 call sites pass `nnei` (sum(sel) from metadata), guaranteeing the tensor width satisfies the compiled graph's dynamic shape constraint.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 7]
```diff
@@ -133,16 +132,29 @@ void DeepSpinPTExpt::init(const std::string& model,
...
+
+  // Read expected nnei (= sum(sel)) — the .pt2 graph has this dimension static.
+  if (metadata.obj_val.count("nnei")) {
+    nnei = metadata["nnei"].as_int();
+  } else {
+    // Fallback: compute from sel array
+    nnei = 0;
+    for (const auto& v : metadata["sel"].as_array()) {
+      nnei += v.as_int();
+    }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:150]: Fixed in 862f5609 — same fix as DeepPotPTExpt, `nnei` is now passed to `createNlistTensor`.

File: source/api_cc/src/commonPTExpt.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/commonPTExpt.h]: Thanks for confirming.

File: source/api_cc/src/DeepPotPTExpt.cc [hunk 8]
```diff
@@ -120,16 +119,29 @@ void DeepPotPTExpt::init(const std::string& model,
...
+  // Read expected nnei (= sum(sel)) — the .pt2 graph has this dimension static.
+  if (metadata.obj_val.count("nnei")) {
+    nnei = metadata["nnei"].as_int();
+  } else {
+    // Fallback: compute from sel array
+    nnei = 0;
+    for (const auto& v : metadata["sel"].as_array()) {
+      nnei += v.as_int();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepPotPTExpt.cc:138]: Fixed in 862f5609 — `nnei` is now used in all `createNlistTensor` calls as the `min_nnei` parameter.

File: source/tests/pt_expt/test_training.py [hunk 9]
```diff
@@ -61,6 +61,17 @@
...
+    "type": "dpa1",
+    "sel": 12,
+    "rcut_smth": 0.50,
+    "rcut": 3.00,
+    "neuron": [8, 16],
+    "axis_neuron": 4,
+    "attn_layer": 2,
+    "precision": "float64",
+    "seed": 1,
+}
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/test_training.py]: Good catch. Removed in 12ad03b9.

File: source/api_cc/src/DeepPotPTExpt.cc [hunk 10]
```diff
@@ -532,10 +546,9 @@ void DeepPotPTExpt::compute(ENERGYVTYPE& ener,
       torch::from_blob(atype_64.data(), {1, nall}, int_options)
           .clone()
           .to(device);
+  // Flatten raw nlist — the .pt2 model sorts by distance on-device.
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepPotPTExpt.cc:564]: Fixed in 862f5609 — all call sites now pass `nnei` (from model metadata) as `min_nnei` to `createNlistTensor`, ensuring the tensor width >= sum(sel).

File: source/api_cc/src/DeepSpinPTExpt.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:151]: Thanks for confirming.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 11]
```diff
@@ -580,10 +599,9 @@ void DeepSpinPTExpt::compute(ENERGYVTYPE& ener,
       torch::from_blob(atype_64.data(), {1, nall}, int_options)
           .clone()
           .to(device);
+  // Flatten raw nlist — the .pt2 model sorts by distance on-device.
   at::Tensor nlist_tensor =
-      buildTypeSortedNlist<double>(nlist_raw, coord_cpy_d, atype_cpy, sel, nloc,
-                                   mixed_types)
-          .to(device);
+      createNlistTensor(nlist_raw).to(torch::kInt64).to(device);
   std::vector<std::int64_t> mapping_64(mapping_vec.begin(), mapping_vec.end());
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:618]: Fixed in 862f5609 — same as DeepPotPTExpt, `createNlistTensor(nlist_raw, nnei)` pads to sum(sel).

File: source/api_cc/src/DeepSpinPTExpt.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc]: Thanks for confirming.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 12]
```diff
@@ -285,32 +297,31 @@ void DeepSpinPTExpt::compute(ENERGYVTYPE& ener,
...
+              .to(device);
     }
-    mapping_tensor =
-        torch::from_blob(mapping.data(), {1, nall_real}, int_option)
-            .clone()
-            .to(device);
+
+    // Flatten raw nlist — the .pt2 model sorts by distance on-device.
+    firstneigh_tensor =
+        createNlistTensor(nlist_data.jlist).to(torch::kInt64).to(device);
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc]: Fixed in 862f5609 — `createNlistTensor(nlist_data.jlist, nnei)` pads to at least sum(sel).

File: deepmd/entrypoints/convert_backend.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/entrypoints/convert_backend.py]: Thanks for confirming.

File: deepmd/pt_expt/utils/serialization.py [hunk 13]
```diff
@@ -336,6 +356,7 @@ def deserialize_to_file(
     model_file: str,
     data: dict,
     model_json_override: dict | None = None,
+    do_atomic_virial: bool = False,
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py:355]: Intentional for performance (2.5x speedup). `dp freeze` passes `do_atomic_virial` explicitly via the model config. `dp convert-backend` has the `--atomic-virial` flag. No existing workflow regresses — callers that need atomic virial already pass it explicitly.

File: source/tests/pt_expt/model/test_spin_ener_model.py [hunk 14]
```diff
@@ -359,6 +359,56 @@ def setUp(self) -> None:
...
+            extend_coord_with_ghosts,
+            normalize_coord,
+        )
+
+        generator = torch.Generator(device="cpu").manual_seed(GLOBAL_SEED)
+        cell = torch.rand([3, 3], dtype=dtype, device="cpu", generator=generator)
+        cell = (cell + cell.T) + 5.0 * torch.eye(3, device="cpu")
+        coord = torch.rand([natoms, 3], dtype=dtype, device="cpu", generator=generator)
+        coord = torch.matmul(coord, cell)
+        atype = torch.tensor([0, 0, 1, 0, 1, 1], dtype=torch.int64)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/model/test_spin_ener_model.py:376]: `natoms` is never changed from the default 6 in this test. The parameter exists for consistency with the test helper pattern, not for arbitrary values. Adding dynamic atype generation would be over-engineering for a fixed test fixture.

File: deepmd/pt_expt/utils/serialization.py [hunk 15]
```diff
@@ -178,24 +185,28 @@ def _make_sample_inputs(
...
         Whether the inputs include an extended_spin tensor.
+    model_nnei : int
+        The model's sum(sel).  Used as the min for the dynamic nnei dim.
     Returns a tuple (not dict) to match positional args of the make_fx
     traced module, whose arg names may have suffixes like ``_1``.
     """
     nframes_dim = torch.export.Dim("nframes", min=1)
     nall_dim = torch.export.Dim("nall", min=1)
     nloc_dim = torch.export.Dim("nloc", min=1)
+    nnei_dim = torch.export.Dim("nnei", min=max(2, model_nnei))
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/utils/serialization.py:206]: Good catch. Fixed in b0096e03 — changed to `max(1, model_nnei)`.

File: source/api_cc/src/DeepPotPTExpt.cc [hunk 16]
```diff
@@ -256,36 +268,32 @@ void DeepPotPTExpt::compute(ENERGYVTYPE& ener,
...
     }
-    mapping_tensor =
-        torch::from_blob(mapping.data(), {1, nall_real}, int_option)
-            .clone()
-            .to(device);
+
+    // Flatten raw nlist — the .pt2 model sorts by distance on-device.
+    firstneigh_tensor =
+        createNlistTensor(nlist_data.jlist).to(torch::kInt64).to(device);
   }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepPotPTExpt.cc:310]: Fixed in 862f5609 — `createNlistTensor(nlist_data.jlist, nnei)` now pads to at least sum(sel).

**Review by iProzd — APPROVED** [review]

**Review by njzjz — APPROVED** [review]
